### PR TITLE
[SPARK-45530][CORE] Use `java.lang.ref.Cleaner` instead of `finalize` for `NioBufferedFileInputStream`

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This pr change to use `java.lang.ref.Cleaner` instead of `finalize()` for `NioBufferedFileInputStream`. They are all protective measures for resource cleanup, but the `finalize()` method has been marked as `deprecated` since Java 9 and will be removed in the future, `java.lang.ref.Cleaner` is the more recommended solution.

https://github.com/openjdk/jdk/blob/dfacda488bfbe2e11e8d607a6d08527710286982/src/java.base/share/classes/java/lang/Object.java#L546-L568

```
     * @deprecated The finalization mechanism is inherently problematic.
     * Finalization can lead to performance issues, deadlocks, and hangs.
     * Errors in finalizers can lead to resource leaks; there is no way to cancel
     * finalization if it is no longer necessary; and no ordering is specified
     * among calls to {@code finalize} methods of different objects.
     * Furthermore, there are no guarantees regarding the timing of finalization.
     * The {@code finalize} method might be called on a finalizable object
     * only after an indefinite delay, if at all.
     *
     * Classes whose instances hold non-heap resources should provide a method
     * to enable explicit release of those resources, and they should also
     * implement {@link AutoCloseable} if appropriate.
     * The {@link java.lang.ref.Cleaner} and {@link java.lang.ref.PhantomReference}
     * provide more flexible and efficient ways to release resources when an object
     * becomes unreachable.
     *
     * @throws Throwable the {@code Exception} raised by this method
     * @see java.lang.ref.WeakReference
     * @see java.lang.ref.PhantomReference
     * @jls 12.6 Finalization of Class Instances
     */
    @Deprecated(since="9")
    protected void finalize() throws Throwable { }
```

### Why are the changes needed?
Clean up deprecated api usage

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by mridulm — COMMENTED** [review]
Do this for other usages of `finalize` as well where we have native resources held ?

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 1]
```diff
@@ -125,13 +132,28 @@ private long skipFromFileChannel(long n) throws IOException {
...
-  protected void finalize() throws IOException {
-    close();
+  private record ResourceCleaner(
+      FileChannel fileChannel,
+      ByteBuffer byteBuffer) implements Runnable {
+    @Override
+    public void run() {
+      try {
+        fileChannel.close();
+        StorageUtils.dispose(byteBuffer);
```
Comment (mridulm) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:153]: This was not done earlier - but do we want to ensure `dispose` is called even if `close` threw an exception ?

**Comment by LuciferYang** [comment]
Yes, for the other `finalize()`, I am also trying to fix them

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java (hunk 1 above)
Comment (LuciferYang) [reply core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:153]: This is a good question, I think even if `fileChannel.close()` throws an exception, we should try to dispose of the byteBuffer, let's put it in the finally block.

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 2]
```diff
@@ -32,6 +34,10 @@
  */
 public final class NioBufferedFileInputStream extends InputStream {
 
+  private static final Cleaner cleaner = Cleaner.create();
+
+  private final Cleaner.Cleanable cleanable;
+
   private static final int DEFAULT_BUFFER_SIZE_BYTES = 8192;
```
Comment (mridulm) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:41]:
```suggestion
  private static final Cleaner cleaner = Cleaner.create();
  private static final int DEFAULT_BUFFER_SIZE_BYTES = 8192;

  private final Cleaner.Cleanable cleanable;
```

**Comment by LuciferYang** [comment]
There are a few more places to do similar work after this pr:

1. RocksDBIterator/LevelDBIterator(SPARK-45533)
2. RemoteBlockPushResolver (SPARK-45534)
3. ThreadWithGarbageCleanup

I have already filed new Jira for Item 1 and 2 for tracking. As for Item 3, since it's in the `hive-thriftserver` module, I have no plans to fix it at the moment.

**Comment by LuciferYang** [comment]
```
[info] *** 1 TEST FAILED ***
[error] Failed tests:
[error] 	org.apache.spark.sql.kafka010.KafkaSourceStressSuite
[warn] In the last 2048 seconds, 6.951 (0.3%) were spent in GC. [Heap: 3.23GB free of 4.00GB, max 4.00GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[error] (sql-kafka-0-10 / Test / test) sbt.TestsFailedException: Tests unsuccessful
[error] Total time: 2049 s (34:09), completed Oct 13, 2023, 4:05:45 PM
```
This is a case that started failing after upgrading to Kafka 3.6, the author is fixing it, it's unrelated to this PR. Could you take another look at this pr If you have time, thanks

File: core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java [hunk 3]
```diff
@@ -32,8 +34,11 @@
  */
 public final class NioBufferedFileInputStream extends InputStream {
 
+  private static final Cleaner CLEANER = Cleaner.create();
```
Comment (LuciferYang) [inline core/src/main/java/org/apache/spark/io/NioBufferedFileInputStream.java:37]: [a9f3025](https://github.com/apache/spark/pull/43333/commits/a9f3025afb2381f573ee0c3e146da9d0b704593b) rename this from `cleaner` to `CLEANER` due to it's static

**Review by mridulm — APPROVED** [review]

**Comment by mridulm** [comment]
Looks good to me, thanks for fixing this @LuciferYang !

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @mridulm ~

**Pull request closed without merging**